File: server/task_service.py
import threading
import time
import uuid
from dataclasses import asdict, dataclass, field
from typing import Any, Dict, List, Optional


@dataclass
class TaskRecord:
    """统一任务记录：下载任务与生图任务共用。"""

    id: str
    kind: str  # download | generation
    name: str
    status: str = "waiting"  # waiting | processing | completed | failed | cancelled
    progress: int = 0
    created_at: float = field(default_factory=time.time)
    error: Optional[str] = None
    meta: Dict[str, Any] = field(default_factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "kind": self.kind,
            "name": self.name,
            "status": self.status,
            "progress": self.progress,
            "createdAt": self.created_at,
            "error": self.error,
            "meta": self.meta,
        }


class TaskService:
    """线程安全的任务注册表（同步方法，便于在下载线程中直接更新）。"""

    def __init__(self) -> None:
        self._tasks: Dict[str, TaskRecord] = {}
        self._lock = threading.Lock()

    def add(self, kind: str, name: str, meta: Optional[Dict[str, Any]] = None) -> TaskRecord:
        record = TaskRecord(
            id=str(uuid.uuid4()),
            kind=kind,
            name=name,
            meta=meta or {},
        )
        with self._lock:
            self._tasks[record.id] = record
        return record

    def get(self, task_id: str) -> Optional[TaskRecord]:
        with self._lock:
            return self._tasks.get(task_id)

    def update(self, task_id: str, **changes: Any) -> Optional[TaskRecord]:
        with self._lock:
            record = self._tasks.get(task_id)
            if record is None:
                return None
            for key, value in changes.items():
                if hasattr(record, key):
                    setattr(record, key, value)
            return record

    def list(self, kind: Optional[str] = None, status: Optional[str] = None) -> List[TaskRecord]:
        with self._lock:
            records = list(self._tasks.values())
        if kind:
            records = [r for r in records if r.kind == kind]
        if status:
            records = [r for r in records if r.status == status]
        return sorted(records, key=lambda r: r.created_at, reverse=True)

    def remove(self, task_id: str) -> bool:
        with self._lock:
            return self._tasks.pop(task_id, None) is not None

    def cancel(self, task_id: str) -> bool:
        with self._lock:
            record = self._tasks.get(task_id)
            if record is None:
                return False
            if record.status in ("completed", "failed", "cancelled"):
                return False
            record.status = "cancelled"
            record.progress = 0
            return True

    def clear_completed(self) -> int:
        with self._lock:
            keys = [
                k
                for k, r in self._tasks.items()
                if r.status in ("completed", "failed", "cancelled")
            ]
            for key in keys:
                del self._tasks[key]
            return len(keys)

File: desktop/src/services/TaskService.ts
import GlobalStateManager from './GlobalState';

export interface QueueTask {
    id: string;
    kind: 'download' | 'generation';
    name: string;
    status: 'waiting' | 'processing' | 'completed' | 'failed' | 'cancelled';
    progress: number;
    createdAt: number;
    error?: string | null;
    meta?: Record<string, any>;
}

/**
 * 任务队列服务：封装 /api/tasks 的 REST 接口与 websocket task_update 订阅。
 */
class TaskService {
    private static instance: TaskService | null = null;
    private ws: WebSocket | null = null;
    private listeners: ((task: QueueTask) => void)[] = [];

    public static getInstance(): TaskService {
        if (!TaskService.instance) {
            TaskService.instance = new TaskService();
        }
        return TaskService.instance;
    }

    private getBaseUrl(): string {
        const rootState = GlobalStateManager.getInstance().getRootState();
        const host = rootState?.host || 'localhost';
        const port = rootState?.port || 7422;
        return `http://${host}:${port}`;
    }

    public async fetchTasks(): Promise<QueueTask[]> {
        const response = await fetch(`${this.getBaseUrl()}/api/tasks`);
        if (!response.ok) {
            throw new Error(`Failed to fetch tasks: ${response.status}`);
        }
        const data = await response.json();
        return data.items || [];
    }

    public async removeTask(id: string): Promise<boolean> {
        const response = await fetch(`${this.getBaseUrl()}/api/tasks/${encodeURIComponent(id)}`, {
            method: 'DELETE'
        });
        const data = await response.json();
        return data.success === true;
    }

    public async cancelTask(id: string): Promise<boolean> {
        const response = await fetch(`${this.getBaseUrl()}/api/tasks/${encodeURIComponent(id)}/cancel`, {
            method: 'POST'
        });
        const data = await response.json();
        return data.success === true;
    }

    public async clearCompleted(): Promise<void> {
        await fetch(`${this.getBaseUrl()}/api/tasks/clear`, {
            method: 'POST'
        });
    }

    /**
     * 通过统一任务队列创建下载任务（HTTP 直链或 HuggingFace 仓库）。
     * @returns 任务 id；失败抛异常。
     */
    public async createDownloadTask(
        kind: string,
        name: string,
        url?: string,
        repo_id?: string
    ): Promise<string> {
        const response = await fetch(`${this.getBaseUrl()}/api/tasks/download`, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json'
            },
            body: JSON.stringify({ kind, name, url, repo_id })
        });
        if (!response.ok) {
            throw new Error(`Failed to create download task: ${response.status}`);
        }
        const data = await response.json();
        if (!data.task_id) {
            throw new Error('Failed to create download task: missing task_id');
        }
        return data.task_id as string;
    }

    public subscribe(callback: (task: QueueTask) => void): () => void {
        this.listeners.push(callback);
        this.ensureSocket();
        return () => {
            this.listeners = this.listeners.filter(listener => listener !== callback);
        };
    }

    private ensureSocket(): void {
        if (this.ws) return;
        const wsUrl = this.getBaseUrl().replace(/^http/, 'ws') + '/ws';
        const socket = new WebSocket(wsUrl);
        socket.onmessage = (event) => {
            try {
                const data = JSON.parse(event.data);
                if (data.type === 'task_update' && data.task) {
                    this.listeners.forEach(listener => listener(data.task as QueueTask));
                }
            } catch (error) {
                console.error('解析任务推送失败:', error);
            }
        };
        socket.onclose = () => {
            this.ws = null;
        };
        this.ws = socket;
    }
}

export default TaskService;
